git-clone/main.tf

```diff
@@ -25,20 +25,82 @@ variable "agent_id" {
   type        = string
 }
 
+variable "git_providers" {
+  type = map(object({
+    provider = string
+  }))
+  description = "The set of git provider by the base url."
+  default = {
+    "https://github.com/" = {
+      provider = "github"
+    },
+    "https://gitlab.com/" = {
+      provider = "gitlab"
+    },
+  }
+  validation {
+    error_message = "Allowed values for provider are \"github\" or \"gitlab\"."
+    condition     = alltrue([for provider in var.git_providers : contains(["github", "gitlab"], provider.provider)])
+  }
+}
+
 locals {
-  clone_path = var.base_dir != "" ? join("/", [var.base_dir, replace(basename(var.url), ".git", "")]) : join("/", ["~", replace(basename(var.url), ".git", "")])
+  # Remove query parameters and fragments from the URL
+  url = replace(replace(var.url, "/\\?.*/", ""), "/#.*/", "")
+
+  # Find the git provider based on the URL and determine the tree path
+  provider_key = try(one([for key in keys(var.git_providers) : key if startswith(local.url, key)]), null)
+  provider     = try(lookup(var.git_providers, local.provider_key).provider, "")
+  tree_path    = local.provider == "gitlab" ? "/-/tree/" : local.provider == "github" ? "/tree/" : ""
+
+  # Remove tree and branch name from the URL
+  clone_url = local.tree_path != "" ? replace(local.url, "/${local.tree_path}.*/", "") : local.url
```

Comment:
Not going to block on this, but is `replace` the best way to handle this rather than some kind of generic matcher (whether that is regex or something else)?  I know in practice it should be fine (like you mentioned, GitHub reserves the word `tree` for example) but even so it feels inflexible and a bit difficult to read.

But, it is something we could iterate on, and maybe we will have to revisit this for the Bitbucket URLs anyway, so like I said no need to block on this.

Reply:
I was thinking that. Until I saw how bitbucket urls are wild for branches with a "/" in them. 

The other idea I am working on is using the git providers api to handle things like pull requests and merge requests.

Reply:
API is an interesting idea, makes a lot of sense.

If we do go with matchers, we could borrow from the routing syntax in go-chi.  It could look something like this from a usage standpoint, not sure how easy it would be to implement in Terraform though...  :sweat_smile:

```
github = [ "{owner}/{repo}/tree/{branch}" ]
gitlab = [ "{owner}/{repo}/-/tree/{branch}" ]
bitbucket = [
  // Each matcher is tried in order.
  "{owner}/{repo}/src/{ignore}?at={branch}"
  "{owner}/{repo}/src/{branch}"
]
```

This could probably be implemented with some regex, but it might be better to actually parse the URLs and then match against the path and the query variables separately, if that can be done.

Reply:
Something else to consider, there are also `blob` URLs that people might try pasting.

```
github = [ "{owner}/{repo}/tree/{branch}" "{owner}/{repo}/blob/{branch}"  ]
```